### DISCUSSION TITLE
[RFC] Global loading replacement

### DISCUSSION THREAD

**Discussion by zombieJ** [ant-design/ant-design#48645] [RFCs - archive]
# 📋 RFC

## Summary

Support customize loading effect to replace with Spin or other loading placeholder if exist:

<img width="45" alt="截屏2024-04-26 11 39 02" src="https://github.com/ant-design/ant-design/assets/5378891/3762cb64-2124-48d9-9b10-0f0dd9eb3794">

<img width="117" alt="截屏2024-04-26 11 39 15" src="https://github.com/ant-design/ant-design/assets/5378891/ea9683fc-a2ad-43d1-9465-b168ae4a68df">

<img width="130" alt="截屏2024-04-26 11 39 47" src="https://github.com/ant-design/ant-design/assets/5378891/fd5dc0a2-830b-4141-b13e-db71813f9e8d">

<img width="177" alt="截屏2024-04-26 11 40 23" src="https://github.com/ant-design/ant-design/assets/5378891/1844cb0e-9811-4e99-bff0-361ca9fa60df">

## Basic example

```tsx
<ConfigProvider renderLoading={() => <div />}  />
```

## Motivation

For the blog [Extends Theme](https://ant-design.antgroup.com/docs/blog/config-provider-style) provide a css way to replace the loading effect but it has the limitation of must be in CSS way and not support some magic effect (e.g. [Dribble: loading search](https://dribbble.com/search/shots/popular/animation?q=loading)).
It should provide a simple way to support the extension. 

## Detailed

`renderLoading` is similar as `renderEmpty` type style:

```tsx
type RenderLoading =(componentName: string, info: { percent?: number }) => ReactNode;
```

When the ConfigProvider config `renderLoading`, it will replace all the loading related content on the UI.

## Drawbacks

N/A

## Adoption strategy

feature in minor

**Comment by afc163** [comment]
```jsx
<ConfigProvider
  spin={{ indicator: ReactNode }}
>
  ...
</ConfigProvider>
```

How about that?

**Reply by zombieJ** [reply]
For Button Select component, it not use Spin. And this RFC want to enhance the `percent` loading logic in each case for more spec case.

**Reply by arvinxx** [reply]
what if I just want to replace the spin other than the button loading?

**Reply by zombieJ** [reply]
> what if I just want to replace the spin other than the button loading?

How about?

```tsx
renderLoading={(component) => component !== 'Spin' ? info.originNode : <div />}
```

**Reply by zombieJ** [reply]
PS: We need wrap on the HappyWork theme which will modify the Button loading status which is not using Spin.

**Reply by afc163** [reply]
`renderLoading={(component) => component !== 'Spin' ? info.originNode : <div />}`

+1

**Comment by arvinxx** [comment]
is `renderLoading` or `loadingRender` ?

**Reply by zombieJ** [reply]
xxxRender is for sub component render. Loading not belong to the ConfigProvider.

**Comment by MadCcc** [comment]
loadingComponent: (props: { loading: boolean, componentName: string }) => ReactElement

loadingComponent={Spin}
loadingComponent={() => <Spin />}

const Spin: FC = (props) => {
  switch(props.componentName) {
    case 'Button':
      return <Spin />
    default:
      return <Default />
  }
}

**Reply by zombieJ** [reply]
```tsx
// 这是合理的
const MySpin = () => {
  const [x, setX] = useState(0);
  useEffect(() => {
    setX(100);
  }, []);

  return <Animation val={x} />
};

loadingComponent={MySpin}
```

**Reply by zombieJ** [reply]
```tsx
const MySpin = () => {
  const [x, setX] = useState(0);
  useEffect(() => {
    setX(100);
  }, []);

  return <Animation val={x} />
};

loadingComponent={(props) => <MySpin {...props} />}
```

**Reply by zombieJ** [reply]
```tsx
const MyButton = () => {
  const [, forceUpdate] = useState();
  
  return <ConfigProvider loadingComponent={(props) => <MySpin {...props} />}>
      <Button onClick={() => forceUpdate({})} />
    </ConfigProvider>

};
```

**Reply by zombieJ** [reply]
```tsx
// render fc
const replacedNode = loadingComponent(props);

// FC
const replacedNode = <LoadingComponent {...props} />
```
